docs/AdvancedUsage.md

```diff
@@ -0,0 +1,801 @@
+    {
+        $this->navBar->click('Sign up');
+        $this->signUp->register([
+            'first_name'            => 'Joe',
+            'last_name'             => 'Jones',
+            'email'                 => 'joe@jones.com',
+            'password'              => '1234',
+            'password_confirmation' => '1234'
+        ]);
+    }
+}
+
+{% endhighlight %}
+
+And for Test classes:
+
+{% highlight php %}
+
+<?php
+namespace Tests\Unit;
+
+class MathTest extends \Codeception\TestCase\Test
+{
+    protected UnitTester $tester;
+    protected Helper\Math $math;
+
+    protected function _inject(\Helper\Math $math)
+    {
+        $this->math = $math;
+    }
+
+    public function testAll()
+    {
+        $this->assertEquals(3, $this->math->add(1, 2));
+        $this->assertEquals(1, $this->math->subtract(3, 2));
+    }
+}
+
+{% endhighlight %}
+
+However, Dependency Injection is not limited to this. It allows you to **inject any class**,
+which can be constructed with arguments known to Codeception.
+
+In order to make auto-wiring work, you will need to implement the `_inject()` method with the list of desired arguments.
+It is important to specify the type of arguments, so Codeception can guess which objects are expected to be received.
+The `_inject()` will only be invoked once, just after creation of the TestCase object (either Cest or Test).
+Dependency Injection will also work in a similar manner for Helper and Actor classes.
+
+Each test of a Cest class can declare its own dependencies and receive them from method arguments:
+
+{% highlight php %}
+
+<?php
+class UserCest
+{
+    function updateUser(\Helper\User $u, \AcceptanceTester $I, \Page\User $userPage)
+    {
+        $user = $u->createDummyUser();
+        $userPage->login($user->getName(), $user->getPassword());
+        $userPage->updateProfile(['name' => 'Bill']);
+        $I->see('Profile was saved');
+        $I->see('Profile of Bill','h1');
+    }
+}
+
+{% endhighlight %}
+
+Moreover, Codeception can resolve dependencies recursively (when `A` depends on `B`, and `B` depends on `C` etc.)
+and handle parameters of primitive types with default values (like `$param = 'default'`).
+Of course, you are not allowed to have *cyclic dependencies*.
+
+## Example Annotation
+
+What if you want to execute the same test scenario with different data? In this case you can inject examples
+as `\Codeception\Example` instances.
+Data is defined via the `@example` annotation, using JSON or Doctrine-style notation (limited to a single line). Doctrine-style:
+
+{% highlight php %}
+
+<?php
+class EndpointCest
+{
+ /**
+  * @example ["/api/", 200]
+  * @example ["/api/protected", 401]
+  * @example ["/api/not-found-url", 404]
+  * @example ["/api/faulty", 500]
+  */
+  public function checkEndpoints(ApiTester $I, \Codeception\Example $example)
+  {
+    $I->sendGet($example[0]);
+    $I->seeResponseCodeIs($example[1]);
+  }
+}
+
+{% endhighlight %}
+
```

Comment:
1. Why do we have 2 copies of all docs files now? If you are renaming, it would be better to replace content of the old page with meta or javascript redirect to the new place.
2. I see that atributes are not documented in AdvancedUsage page.